File: backend/py_template/devdonalds.py
from dataclasses import dataclass
from typing import List, Dict, Union
from flask import Flask, request, jsonify
import re

# ==== Type Definitions, feel free to add or modify ===========================
@dataclass
class CookbookEntry:
	name: str

@dataclass
class RequiredItem():
	name: str
	quantity: int

@dataclass
class Recipe(CookbookEntry):
	required_items: List[RequiredItem]

@dataclass
class Ingredient(CookbookEntry):
	cook_time: int


# =============================================================================
# ==== HTTP Endpoint Stubs ====================================================
# =============================================================================
app = Flask(__name__)

# Store your recipes here!
cookbook = {}

# Task 1 helper (don't touch)
@app.route("/parse", methods=['POST'])
def parse():
	data = request.get_json()
	recipe_name = data.get('input', '')
	parsed_name = parse_handwriting(recipe_name)
	if parsed_name is None:
		return 'Invalid recipe name', 400
	return jsonify({'msg': parsed_name}), 200

# [TASK 1] ====================================================================
# Takes in a recipeName and returns it in a form that 
def parse_handwriting(recipeName: str) -> Union[str | None]:
	goodRecipeName = ''
	t = False
	for cha in recipeName:
		if len(goodRecipeName) != 0:
			lastCha = goodRecipeName[-1]
		else:
			lastCha = 'anything'
		if cha == '-' or cha == '_' or cha == ' ':
			if lastCha != ' ':						# So that there is just one space
				goodRecipeName += ' '
		elif cha.isalpha():
			if t == False:							# Uppercase the first letter
				goodRecipeName += cha.upper()
				t = True
			elif lastCha == ' ':
				goodRecipeName += cha.upper()		# Uppercase the first letter of a word
			else:
				goodRecipeName += cha.lower()		# Lowercase the rest of the letters
    
	if len(goodRecipeName) == 0:
		return None
	recipeName = goodRecipeName
	return recipeName


# [TASK 2] ====================================================================
# Endpoint that adds a CookbookEntry to your magical cookbook
@app.route('/entry', methods=['POST'])
def create_entry():
	data = request.get_json()
	entries_type = data.get('type', '')
	entries_name = data.get('name', '')
	if entries_type != 'recipe' and entries_type != 'ingredient':
		return 'Invalid type', 400
	if entries_name in cookbook:
		return 'Entry names already exists', 400
	if entries_type == 'ingredient':
		cookTime = int(data.get('cookTime', ''))
		if cookTime <= 0:
			return 'Invalid cookTime', 400
	cookbook[entries_name] = data
	return 'implemented successfully', 200


# [TASK 3] ====================================================================
# Endpoint that returns a summary of a recipe that corresponds to a query name   
@app.route('/summary', methods=['GET'])
def summary():
    recipe_name = request.args.get('name', '')
    if recipe_name not in cookbook or cookbook[recipe_name]["type"] != "recipe":
        return 'Invalid recipe', 400
    
    total_cooktime = 0
    ingredients = {}
    
    def process_requiredItems(item_name, quantity):
        nonlocal total_cooktime
        if item_name not in cookbook:
            return False
        item = cookbook[item_name]
        if item["type"] == "ingredient":
            total_cooktime += item["cookTime"] * quantity
            ingredients[item_name] = ingredients.get(item_name, 0) + quantity
        elif item["type"] == "recipe":
            for sub_item in item["requiredItems"]:
                if not process_requiredItems(sub_item["name"], sub_item["quantity"] * quantity):
                    return False
        return True
    
    for required_item in cookbook[recipe_name]["requiredItems"]:
        if not process_requiredItems(required_item["name"], required_item["quantity"]):
            return 'Invalid recipe', 400
    
    return jsonify({
        "name": recipe_name,
        "cookTime": total_cooktime,
        "ingredients": [{"name": name, "quantity": qty} for name, qty in ingredients.items()]
    })



# =============================================================================
# ==== DO NOT TOUCH ===========================================================
# =============================================================================

if __name__ == '__main__':
	app.run(debug=True, port=8080)

File: backend/ts_template/devdonalds.ts
import express, { Request, Response } from "express";

// ==== Type Definitions, feel free to add or modify ==========================
interface cookbookEntry {
  name: string;
  type: string;
}

interface requiredItem {
  name: string;
  quantity: number;
}

interface recipe extends cookbookEntry {
  requiredItems: requiredItem[];
}

interface ingredient extends cookbookEntry {
  cookTime: number;
}

// =============================================================================
// ==== HTTP Endpoint Stubs ====================================================
// =============================================================================
const app = express();
app.use(express.json());

// Store your recipes here!
const cookbook = new Map<string, cookbookEntry>();

// Task 1 helper (don't touch)
app.post("/parse", (req:Request, res:Response) => {
  const { input } = req.body;

  const parsed_string = parse_handwriting(input)
  if (parsed_string == null) {
    res.status(400).send("this string is cooked");
    return;
  } 
  res.json({ msg: parsed_string });
  return;
  
});

// [TASK 1] ====================================================================
// Takes in a recipeName and returns it in a form that 
const parse_handwriting = (recipeName: string): string | null => {
  if (!recipeName.match(/[a-zA-Z]/)) {
      return null;
  }
  recipeName = recipeName.replace(/[-_]+/g, ' ');
  recipeName = recipeName.replace(/[^a-zA-Z\s]/g, '');
  let words = recipeName.split(" ");
  words = words.filter(word => word !== "");
  
  if (words.length === 0) {
      return null;
  }
  let finalWords = [];
  for (let word of words) {
      let newWord = word.toLowerCase();
      newWord = newWord.charAt(0).toUpperCase() + newWord.slice(1);
      finalWords.push(newWord);
  }
  return finalWords.join(" ");
}

// [TASK 2] ====================================================================
// Endpoint that adds a CookbookEntry to your magical cookbook
app.post("/entry", (req:Request, res:Response) => {
  const entry = req.body;
  if (entry.type !== "recipe" && entry.type !== "ingredient") {
    return res.status(400).send("Invalid type");
  }
  if (cookbook.has(entry.name)) {
    return res.status(400).send("Entry name must be unique");
  }

  if (entry.type === "ingredient") {
    const ingredientEntry = entry as ingredient;
    if (typeof ingredientEntry.cookTime !== "number" || ingredientEntry.cookTime < 0) {
      return res.status(400).send("cookTime must be greater than or equal to 0");
    }
  } else {
    const recipeEntry = entry as recipe;
    const itemNames = new Set<string>();
    for (const item of recipeEntry.requiredItems) {
      if (itemNames.has(item.name)) {
        return res.status(400).send("Recipe requiredItems must have unique names");
      }
      if (typeof item.quantity !== "number" || item.quantity <= 0) {
        return res.status(400).send("Item quantity must be a positive number");
      }
      itemNames.add(item.name);
    }
  }
  cookbook.set(entry.name, entry);
  return res.status(200).send();

});

// [TASK 3] ====================================================================
// Endpoint that returns a summary of a recipe that corresponds to a query name
app.get("/summary", (req: Request, res: Response) => { 
  const name = req.query.name as string;
  const entry = cookbook.get(name);
  if (!entry) {
    return res.status(400).send("Can't find that recipe!");
  }
  if (entry.type !== "recipe") {
    return res.status(400).send("That's not a recipe!");
  }
  let totalCookTime = 0;
  const ingredients: requiredItem[] = [];
  const seenIngredients = new Map<string, number>();

  const processRecipe = (recipe: recipe, multiplier: number) => {
    for (const item of recipe.requiredItems) {
      const subEntry = cookbook.get(item.name);
      if (!subEntry) {
        throw new Error("Missing ingredient!");
      }

      if (subEntry.type === "ingredient") {
        const ing = subEntry as ingredient;
        totalCookTime += ing.cookTime * item.quantity * multiplier;
        
        const currentQty = seenIngredients.get(item.name) || 0;
        seenIngredients.set(item.name, currentQty + (item.quantity * multiplier));
      } else {
        processRecipe(subEntry as recipe, item.quantity * multiplier);
      }
    }
  };

  try {
    processRecipe(entry as recipe, 1);
    seenIngredients.forEach((qty, name) => {
      ingredients.push({ name, quantity: qty });
    });
    ingredients.sort((a, b) => a.name.localeCompare(b.name));

    return res.status(200).json({
      name: name,
      cookTime: totalCookTime,
      ingredients: ingredients
    });

  } catch (error) {
    return res.status(400).send(error.message);
  }
});

// =============================================================================
// ==== DO NOT TOUCH ===========================================================
// =============================================================================
const port = 8080;
app.listen(port, () => {
  console.log(`Running on: http://127.0.0.1:8080`);
});
